src/NodeServices/PythonServices.cs

```diff
@@ -0,0 +1,326 @@
+                        eventHandlerName = nameof(HandleEvaluationBegin);
+                    }
+                    else if (eventInfo.Name.Equals("EvaluationEnd"))
+                    {
+                        eventHandlerName = nameof(HandleEvaluationEnd);
+                    }
+                    else
+                    {
+                        continue;
+                    }
+
+                    MethodInfo handlerInfo = typeof(PythonEngineProxy).GetMethod(eventHandlerName, BindingFlags.NonPublic | BindingFlags.Instance);
+                    var handler = Delegate.CreateDelegate(eventInfo.EventHandlerType, this, handlerInfo);
+                    eventInfo.RemoveEventHandler(this, handler);
+                }
+            }
+            catch
+            { }
+        }
+
+        /// <summary>
+        /// Add an event handler before the Python evaluation begins
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationBegin(Action<string, IList, Action<string, object>> callback)
+        {
+
+            OnBeginEvaluation = callback;
+            AddEventHandler("EvaluationStarted", nameof(HandleEvaluationBegin));
+        }
+
+        /// <summary>
+        /// Add an event handler after the Python evaluation ends
+        /// </summary>
+        /// <param name="callback"></param>
+        public void OnEvaluationEnd(Action<string, IList> callback)
+        {
+            OnEndEvaluation = callback;
+            AddEventHandler("EvaluationFinished", nameof(HandleEvaluationEnd));
+        }
+
+        /// <summary>
+        /// Gets the input Marshaler from the target Python Engine
+        /// </summary>
+        /// <returns>DataMarshaler as object</returns>
+        public object GetInputMarshaler()
+        {
+            try
+            {
+                return EngineType.GetProperty(PythonEngineManager.PythonInputMarshalerProperty).GetValue(null);
+            }
+            catch
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Gets the output Marshaler from the target Python Engine
+        /// </summary>
+        /// <returns>DataMarshaler as object</returns>
+        public object GetOutputMarshaler()
+        {
+            try
+            {
+                return EngineType.GetProperty(PythonEngineManager.PythonOutputMarshalerProperty).GetValue(null);
+            }
+            catch
+            {
+                return null;
+            }
+        }
+
+        private void HandleEvaluationBegin(string code, IList bindingValues, Action<string, object> scopeFn)
+        {
+            OnBeginEvaluation?.Invoke(code, bindingValues, scopeFn);
+        }
+
+        private void HandleEvaluationEnd(string code, IList bindingValues)
+        {
+            OnEndEvaluation?.Invoke(code, bindingValues);
+        }
+
+
+        private void AddEventHandler(string targetEventName, string handlerName)
```

Comment:
this acts dynamically because the underlying type doesn't have an interface we can use (or shared base class)?